pydantic/main.py

```diff
@@ -602,7 +611,7 @@ def validate_model(  # noqa: C901 (ignore complexity)
         elif check_extra:
             names_used.add(field.name if using_name else field.alias)
 
-        v_, errors_ = field.validate(value, values, loc=field.alias, cls=model.__class__)  # type: ignore
+        v_, errors_ = field.validate(value, values, loc=field.alias, cls=cls or model.__class__)  # type: ignore
```

Comment:
not sure why this is required. Can you explain?

Reply:
Without this change, this test would fail: the ```assert cls is``` line

```def test_classmethod():
    @dataclass
    class MyDataclass:
        a: str

        @validator('a')
        def check_a(cls, v):
            assert cls is MyDataclass and is_dataclass(MyDataclass)
            return v

    m = MyDataclass(a='this is foobar good')
    assert m.a == 'this is foobar good'
    m.check_a('x')
```

We would expect the cls attribute to the validator method (check_a in this case) to be the Dataclass, not the `__pydantic_model__`. However, without this change, it would be the `__pydantic_model__` BaseModel.

The `cls` kwarg to `field.validate` in the line you referred to is eventually passed to the validator classmethod as the cls arg. The `cls` arg should refer to the original class the validator is defined on, which can either be a BaseModel or a Dataclass.

`model.__class__` which is what was previously given works for `BaseModel`. However for dataclass, this would be the `__pydantic_model__`  class instead of the original dataclass.

The Dataclass now gives the original dataclass as the `cls`  kwarg to `validate_model`  thus the dataclass is what is eventually passed to the validator, not the `__pydantic_model__` BaseModel.

Hope it's clear.

Reply:
makes sense.